include/zephyr/kernel/smp.h

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2023 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_KERNEL_SMP_H_
+#define ZEPHYR_INCLUDE_KERNEL_SMP_H_
+
+typedef void (*smp_custom_init_fn)(void *arg);
+
+/**
+ * @brief Start one CPU.
+ *
+ * This starts (powers up/starts up) a CPU that has stopped
+ * or powered down. After the CPU is started, the scheduler
+ * will run, and the CPU will participate in execution of
+ * threads.
+ *
+ * @note This unconditionally starts a CPU regardless of
+ *       its current state. Depending on the architecture,
+ *       it may have no effect or very bad side effect.
+ *       Use this with caution.
+ *
+ * @warning Parallel calls to this function will result
+ *          in undesired behavior. Please make sure only
+ *          one call to this function at a time.
+ *
+ * @param id ID of target CPU.
+ */
+void k_smp_cpu_start(int id);
+
+/**
+ * @brief Start one CPU with additional initialization callback.
+ *
+ * This starts (powers up/starts up) a CPU that has stopped
+ * or powered down. After the CPU is started, the provided
+ * function (@a fn) will be called to perform any additional
+ * initialization (e.g. setting up data structures). After
+ * that, the scheduler will run, and the CPU will participate
+ * in execution of threads.
+ *
+ * @note This unconditionally starts a CPU regardless of
+ *       its current state. Depending on the architecture,
+ *       it may have no effect or very bad side effect.
+ *       Use this with caution.
```

Comment:
Nitpick, but a comment saying effectively "what this function does is going to depend on the SOC" is probably a good clue that it should still be a SOC-provided API :)

Can we tighten up the spec a little?   Maybe document this like "This function must not be called on currently-executing CPUs, the device must have reached its idle/off state.  Detection of that state in the general case must be provided by the platform layers."

Reply:
Updated.